### ISSUE TITLE
[Testing]: Comprehensive testing and validation suite for MCP sampling

### ISSUE THREAD

**Issue by kesslerio** [kesslerio/vibe-check-mcp#197]
### Context
PR #190 implemented basic MCP sampling with minimal testing (21 unit tests). This issue addresses the need for comprehensive testing to ensure production readiness, edge case handling, and performance validation.

### Related Issues
- Parent Issue: #189 (Dynamic Response Generation vision)
- Initial Implementation: PR #190 (Basic implementation with minimal tests)
- Security Issues: #194 (Must be fixed before comprehensive testing)

### 1. Novel Query Testing 🆕
Test system behavior with queries that break current assumptions:

- [ ] **Edge Case Queries**
  ```python
  test_cases = [
      # Extremely specific technical queries
      "Should I use Rust's Arc<Mutex<T>> or Rc<RefCell<T>> for my single-threaded event loop?",
      
      # Multi-domain queries
      "How do I integrate Terraform with our custom Kubernetes operator while maintaining GitOps principles?",
      
      # Ambiguous queries requiring clarification
      "Is it better?",
      
      # Queries with contradictory requirements
      "I need a stateless system that maintains user sessions",
      
      # Non-English or mixed language queries
      "Comment implémenter a factory pattern en Python?",
      
      # Queries with code injection attempts
      "Help with: '); DROP TABLE users; --",
      
      # Extremely long queries (>10k characters)
      "[10,000 character technical specification]..."
  ]
  ```

- [ ] **Query Categories to Test**
  - Never-before-seen technologies
  - Highly specific version combinations
  - Abstract architectural philosophy questions
  - Queries mixing multiple unrelated domains
  - Time-sensitive queries ("latest in 2024")
  - Queries requiring external knowledge

- [ ] **Failure Mode Testing**
  - Document how system fails on impossible queries
  - Verify graceful degradation
  - Test fallback mechanisms
  - Ensure no crashes or hangs

### 2. Response Consistency Validation ✅
Ensure responses maintain quality and consistency:

- [ ] **Consistency Test Suite**
  ```python
  class ConsistencyValidator:
      def test_persona_consistency(self, responses: List[str]) -> float:
          """Verify vibe_check_mentor voice maintained"""
          
      def test_technical_accuracy(self, response: str, ground_truth: str) -> float:
          """Check technical correctness"""
          
      def test_advice_consistency(self, similar_queries: List[str]) -> float:
          """Ensure similar queries get compatible advice"""
  ```

- [ ] **Test Scenarios**
  - Same query asked 10 times → similar responses
  - Slight query variations → consistent core advice
  - Different phrasings of same problem → aligned solutions
  - Persona voice across different topics
  - No contradictory advice in related queries

- [ ] **Quality Metrics**
  - Response relevance score (0-100)
  - Technical accuracy rate
  - Persona consistency score
  - Hallucination detection rate
  - Actionability assessment

### 3. Performance Comparison Testing 📊
Compare static vs dynamic response systems:

- [ ] **Benchmark Suite**
  ```python
  @pytest.mark.benchmark
  class TestPerformanceComparison:
      def test_static_response_latency(self, benchmark):
          # Measure static response time
          
      def test_dynamic_response_latency(self, benchmark):
          # Measure dynamic generation time
          
      def test_hybrid_routing_overhead(self, benchmark):
          # Measure routing decision cost
          
      def test_cache_effectiveness(self, benchmark):
          # Measure cache hit impact
  ```

- [ ] **Metrics to Compare**
  - Response time (P50, P95, P99)
  - Memory usage
  - CPU utilization
  - Token consumption
  - Cache hit rates
  - Quality scores

- [ ] **Load Testing Scenarios**
  - Single user, sequential queries
  - 10 concurrent users
  - 100 concurrent users
  - Burst traffic patterns
  - Sustained high load
  - Mixed query complexity

### 4. User Acceptance Testing 👥
Validate with real-world usage patterns:

- [ ] **Test User Groups**
  - Junior developers (need detailed guidance)
  - Senior engineers (need strategic advice)
  - DevOps engineers (infrastructure queries)
  - Team leads (architecture decisions)
  - Non-technical stakeholders (high-level explanations)

- [ ] **UAT Scenarios**
  ```python
  uat_scenarios = {
      "junior_dev": [
          "How do I set up ESLint in my React project?",
          "What's the difference between let and const?",
          "Should I use Redux or Context API?"
      ],
      "senior_eng": [
          "Microservices vs modular monolith for our scale?",
          "Event sourcing trade-offs for financial system?",
          "How to migrate from REST to GraphQL incrementally?"
      ]
  }
  ```

- [ ] **Acceptance Criteria**
  - 85%+ user satisfaction rate
  - Advice deemed actionable by users
  - No critical misinformation
  - Response time acceptable to users
  - Clear and understandable responses

### 5. Edge Case Test Suite 🔧
Comprehensive edge case coverage:

- [ ] **Input Edge Cases**
  - Empty queries
  - Single character queries
  - Only punctuation/symbols
  - Extremely long queries
  - Binary/hex data
  - Malformed JSON/code in queries

- [ ] **Context Edge Cases**
  - No workspace context available
  - Corrupted workspace data
  - Massive workspace context (>1MB)
  - Circular dependencies in code
  - Non-UTF8 file content
  - Symbolic links and special files

- [ ] **System Edge Cases**
  - MCP client unavailable
  - Network timeouts mid-response
  - Memory pressure scenarios
  - Disk full conditions
  - Clock skew/time changes
  - Process interruption/signals

- [ ] **Concurrency Edge Cases**
  - Simultaneous identical queries
  - Cache race conditions
  - Circuit breaker state transitions
  - Resource contention
  - Deadlock scenarios

### 6. Security Testing 🔒
Validate security hardening:

- [ ] **Injection Testing**
  - Prompt injection attempts
  - Path traversal in file contexts
  - Command injection via queries
  - Template injection attacks
  - Unicode/encoding attacks

- [ ] **Resource Exhaustion**
  - Memory exhaustion attempts
  - CPU spinning queries
  - Cache poisoning
  - Rate limit bypass attempts
  - Amplification attacks

- [ ] **Data Leakage**
  - Sensitive file access attempts
  - Cross-user data leakage
  - Model extraction attacks
  - Training data extraction
  - System prompt extraction

### 7. Regression Test Suite 📝
Prevent functionality degradation:

- [ ] **Core Functionality Tests**
  - All original static responses still work
  - Workspace awareness maintains accuracy
  - Intent classification remains reliable
  - Entity extraction functions correctly
  - Circuit breaker operates properly

- [ ] **Integration Tests**
  - MCP server integration
  - File system operations
  - Cache operations
  - Logging and telemetry
  - Error handling paths

- [ ] **Automated Test Pipeline**
  ```yaml
  regression_tests:
    - unit_tests: 100+ tests
    - integration_tests: 50+ tests
    - performance_tests: 20+ benchmarks
    - security_tests: 30+ scenarios
    - smoke_tests: 10 critical paths
  ```

### Test Data Management
- [ ] Create test query corpus (1000+ queries)
- [ ] Build ground truth response set
- [ ] Generate synthetic workspace data
- [ ] Create malicious input dataset
- [ ] Maintain test user personas

### Test Automation
- [ ] CI/CD pipeline integration
- [ ] Automated performance regression detection
- [ ] Nightly security scanning
- [ ] Weekly load testing runs
- [ ] Monthly UAT cycles

### Test Reporting
- [ ] Test coverage reports (target: >90%)
- [ ] Performance trend dashboards
- [ ] Security scan results
- [ ] Quality score tracking
- [ ] Regression detection alerts

## Acceptance Criteria

- [ ] Test coverage > 90% for critical paths
- [ ] All edge cases have explicit tests
- [ ] Performance benchmarks established
- [ ] Security tests passing (post #194 fixes)
- [ ] UAT satisfaction > 85%
- [ ] No P0/P1 bugs in regression suite
- [ ] Test execution time < 10 minutes for CI
- [ ] Comprehensive test documentation

## Success Metrics

- Test Coverage: > 90% for src/vibe_check/mentor/
- Bug Detection Rate: Find 20+ issues before production
- Performance Regression Detection: < 24 hours
- Test Reliability: < 1% flaky tests
- Security Coverage: 100% of OWASP top 10 applicable items

## Definition of Done

- [ ] All test categories implemented
- [ ] Test data corpus created and validated
- [ ] CI/CD pipeline running all tests
- [ ] Performance baselines established
- [ ] Security testing complete and passing
- [ ] UAT completed with documented results
- [ ] Test documentation and runbooks created
- [ ] Team trained on test maintenance
- [ ] Monitoring alerts configured for test failures
